Source/WTF/wtf/PreciseSum.h

```diff
@@ -32,35 +32,91 @@ namespace WTF {
+        Vector<int64_t> &&chunk, const int addsUntilPropagate, const int64_t inf,
+        const int64_t nan, const size_t sizeCount, const bool hasPosNumber
+    );
     ~SmallAccumulator() = default;
 
     Vector<int64_t> chunk; // Chunks making up small accumulator
     int addsUntilPropagate; // Number of remaining adds before carry
     int64_t inf; // If non-zero, +Inf, -Inf, or NaN
     int64_t nan; // If non-zero, a NaN value with payload
+    size_t sizeCount; // number of added values
+    bool hasPosNumber; // check if added values have at least one positive number
+
+    int carryPropagate();
+    void addInfNan(int64_t ivalue);
+    inline void add1NoCarry(double value);
+    ALWAYS_INLINE void incrementWhenValueAdded(double value);
+};
+
+struct LargeAccumulator final {
+    Vector<uint64_t> chunk; // Chunks making up large accumulator
+    Vector<int_least16_t> count; // Counts of # adds remaining for chunks, or -1 if not used yet or special
+    Vector<uint64_t> chunksUsed; // Bits indicate chunks in use
+    uint64_t usedUsed; // Bits indicate chunk_used entries not 0
+    SmallAccumulator sacc; // The small accumulator to condense into
+
+    explicit LargeAccumulator();
+    ~LargeAccumulator() = default;
+
+    void addLchunkToSmall(int_fast16_t ix);
+    void largeAddValueInfNan(int_fast16_t ix, uint64_t uintv);
+    void transferToSmall();
+};
+
+class XsumInterface {
+public:
+    virtual ~XsumInterface() = default;
+
+    virtual void addList(const std::span<const double> vec) = 0;
+    virtual void add(double value) = 0;
+    virtual double compute() = 0;
+};
+
+class XsumSmall final : public XsumInterface {
+public:
+    explicit XsumSmall();
+    explicit XsumSmall(SmallAccumulator sacc);
+    ~XsumSmall() = default;
+
+    void addList(const std::span<const double> vec) override;
+    void add(double value) override;
+    double compute() override;
+
+private:
+    SmallAccumulator m_smallAccumulator;
+};
+
+class XsumLarge final : public XsumInterface {
+public:
+    explicit XsumLarge();
+    ~XsumLarge() = default;
+
+    void addList(const std::span<const double> vec) override;
+    void add(double value) override;
+    double compute() override;
+
+private:
+    LargeAccumulator m_largeAccumulator;
 };
 
 } // namespace Xsum
 
 class PreciseSum final {
 public:
     explicit PreciseSum();
+    explicit PreciseSum(uint64_t expectedInputSize);
     ~PreciseSum() = default;
 
     WTF_EXPORT_PRIVATE void addList(const std::span<const double> vec);
     WTF_EXPORT_PRIVATE void add(double value);
     WTF_EXPORT_PRIVATE double compute();
 
 private:
-    Xsum::SmallAccumulator m_smallAccumulator;
-    size_t m_sizeCount;
-    bool m_hasPosNumber;
-
-    void xsumSmallAddInfNan(int64_t ivalue);
-    inline void xsumAdd1NoCarry(double value);
-    int xsumCarryPropagate();
-    ALWAYS_INLINE void incrementWhenValueAdded(double value);
+    std::unique_ptr<Xsum::XsumInterface> m_accumulator;
```

Comment:
I don't feel like this is great for perf as it is always including heap allocation & frequent virtual function calls. Given that there is only two modes, switching them via enum would be better for perf. (And in particular, because this patch is for perf, we should do it).